src/serenity/config.hpp

```diff
@@ -11,140 +12,133 @@
-  }
+    Result<T> result = None();
 
-  // -- unsafe getters --
+    // Get item from items map.
+    Option<SerenityConfig::CfgVariant> variantResult = getItem(key);
 
-  /**
-   * Unsafe getter for string
-   */
-  std::string getS(std::string key) {
-    return boost::get<std::string>(this->fields[key]);
-  }
+    if (variantResult.isSome()) {
+      // When item is found, try to parse it to the specified T type.
+      try {
+        result = boost::get<T>(variantResult.get());
+      } catch (std::exception& e) {
+        LOG(ERROR) << "Failed to parse " << key
+                   << " field: " << e.what();
+        result = Result<T>::error(e.what());
+      }
+    }
 
-  /**
-   * Unsafe getter for int64_t
-   */
-  int64_t getI64(std::string key) {
-    return boost::get<int64_t>(this->fields[key]);
+    return result;
   }
 
   /**
-   * Unsafe getter for uint64_t
+   * Templated, safe getter for item in config. Sets default value in case of
+   * error or none.
    */
-  uint64_t getU64(std::string key) {
-    return boost::get<uint64_t>(this->fields[key]);
-  }
+  template <typename T>
+  const T item(const std::string& key, T defaultValue) const {
+    static_assert(std::is_same<T, bool>()
+                  || std::is_same<T, int64_t>()
+                  || std::is_same<T, double_t>()
+                  || std::is_same<T, std::string>(),
+                  "T must be one of the types stored in CfgVariant (bool, "
+                    "int64_t, double_t, string)");
+    T result = defaultValue;
 
-  /**
-   * Unsafe getter for double_t
-   */
-  double_t getD(std::string key) {
-    return boost::get<double_t>(this->fields[key]);
-  }
+    // Get item from items map.
+    Option<SerenityConfig::CfgVariant> variantResult = getItem(key);
 
-  /**
-   * Unsafe getter for bool
-   */
-  bool getB(std::string key) {
-    return boost::get<bool>(this->fields[key]);
-  }
-
-  // -- setters --
-
-  /**
-   * Sets char* config value.
-   */
-  void set(std::string key, char* value) {
-    this->setVariant(key, (std::string)value);
-  }
-
-  /**
-   * Sets string config value.
-   */
-  void set(std::string key, std::string value) {
-    this->setVariant(key, value);
-  }
+    if (variantResult.isSome()) {
+      // When item is found, try to parse it to the specified T type.
+      try {
+        result = boost::get<T>(variantResult.get());
+      } catch (std::exception& e) {
+        LOG(ERROR) << "Failed to parse " << key << " to type "
+                   << typeid(T).name() << ". Field: " << e.what();
+      }
+    }
 
-  /**
-   * Sets bool config value.
-   */
-  void set(std::string key, bool value) {
-    this->setVariant(key, value);
+    return result;
   }
 
   /**
-   * Sets uint64_t config value.
+   * Gets config section.
+   * In case there is not one, create empty section.
    */
-  void set(std::string key, uint64_t value) {
-    this->setVariant(key, value);
+  SerenityConfig& operator[](const std::string& key) {
```

Comment:
operator[] does not tell user what it is really is for... I would rather explicity call it getSection.

How about removing operator[] and rename it to `Try<SerenityConfig&> getSection?`. We could make SerenityConfig immutable.